utils/dyn-config.c

```diff
@@ -0,0 +1,597 @@
+  INIT_LIST_HEAD(&option->list);
+
+  list_add_tail(&gb_options, &option->list);
+  return option;
+
+freeOption:
+  GB_FREE(option);
+  return NULL;
+}
+
+static void
+glusterBlockParseOption(char **cur, const char *end)
+{
+  struct gbConfOption *option;
+  gbOptionType type;
+  char *p = *cur, *q = *cur, *r, *s;
+
+  while (isblank(*p)) {
+    p++;
+  }
+
+  GB_TO_LINE_END(q, end);
+  *q = '\0';
+  *cur = q + 1;
+
+  /* parse the boolean type option */
+  s = r = strchr(p, '=');
+  if (!r) {
+    /* boolean type option at file end or line end */
+    r = p;
+    while (!isblank(*r) && r < q) {
+      r++;
+    }
+    *r = '\0';
+    option = glusterBlockGetOption(p);
+    if (!option) {
+      option = glusterBlockRegisterOption(p, GB_OPT_BOOL);
+    }
+
+    if (option) {
+      option->optBool = true;
+    }
+
+    return;
+  }
+  /* skip character '='  */
+  s++;
+  r--;
+  while (isblank(*r)) {
+    r--;
+  }
+  r++;
+  *r = '\0';
+
+  option = glusterBlockGetOption(p);
+  if (!option) {
+    r = s;
+    while (isblank(*r)) {
+      r++;
+    }
+    LOG("dyncfg", GB_LOG_INFO, "option type s:'%s', r:'%s' not supported!\n", s, r);
+
+    if (isdigit(*r)) {
+      type = GB_OPT_INT;
+    } else {
+      type = GB_OPT_STR;
+    }
+
+    option = glusterBlockRegisterOption(p, type);
+    if (!option) {
+      return;
+    }
+  }
+
+  /* parse the int/string type options */
+  switch (option->type) {
+  case GB_OPT_INT:
+    while (!isdigit(*s)) {
+      s++;
+    }
+    r = s;
+    while (isdigit(*r)) {
+      r++;
+    }
+    *r= '\0';
+
+    option->optInt = atoi(s);
+    break;
+  case GB_OPT_STR:
+    while (isblank(*s)) {
+      s++;
+    }
+    /* skip first " or ' if exist */
+    if (*s == '"' || *s == '\'') {
+      s++;
+    }
+    r = q - 1;
+    while (isblank(*r)) {
+      r--;
+    }
+    /* skip last " or ' if exist */
+    if (*r == '"' || *r == '\'') {
+      *r = '\0';
+    }
+
+    /* free if this is reconfig */
+    if (option->optStr) {
+      GB_FREE(option->optStr);
+    }
+    GB_STRDUP(option->optStr, s);
+    break;
+  default:
+    LOG("dyncfg", GB_LOG_ERROR,
+        "option type %d not supported!\n", option->type);
+    break;
+  }
+}
+
+static void
+glusterBlockParseOptions(gbConfig *cfg, char *buf, int len, bool reloading)
+{
+  char *cur = buf, *end = buf + len;
+
+  while (cur < end) {
+    /* skip blanks lines */
+    GB_SKIP_BLANK_LINES(cur, end);
+
+    /* skip comments with '#' */
+    GB_SKIP_COMMENT_LINES(cur, end);
+
+    if (cur >= end) {
+      break;
+    }
+
+    if (!isalpha(*cur)) {
+      continue;
+    }
+
+    /* parse the options from config file to gb_options[] */
+    glusterBlockParseOption(&cur, end);
+  }
+
+  /* parse the options from gb_options[] to struct gbConfig */
+  glusterBlockConfSetOptions(cfg, reloading);
+}
+
+static int
+glusterBlockLoadConfig(gbConfig *cfg, bool reloading)
+{
+  int ret = -1;
+  int fd, len;
+  char *buf;
+
+  buf = CALLOC(GB_MAX_CFG_FILE_SIZE);
```

Comment:
Use GB_ALLOC() instead of CALLOC()

GB_ALLOC() will internally uses calloc()

Reply:
This won't work, the buf here is one 'char *' pointer, not struct type pointer.

Sep 20 21:35:59 rhel3 gluster-blockd[7532]: *** Error in `/usr/local/sbin/gluster-blockd': free(): invalid next size (fast): 0x0000000000db9070 ***

Reply:
What ?

Did you use:
```
if (GB_ALLOC_N(buf, GB_MAX_CFG_FILE_SIZE) < 0) {
...
}
```

Reply:
Okay, will use this. :-)